### ISSUE TITLE
Panic when getting module gradients with autodiff backend

### ISSUE THREAD

**Issue by lucasmdjl** [tracel-ai/burn#3969]
**Describe the bug**
When trying to get gradients for a specific module with `GradientsParams::get` and an `AutodiffBackend`, the code panics because it tries to unwrap a `None` `Option`. After digging a bit into the code, I've seen this is because `GradientsParams::from_grads` uses `B::InnerBackend` to register. Of course, I can use the inner backend directly when calling `GradiensParams::get`, but this isn't intuitive (at least to me) and there's no indication that it should be called like that unless you dig into the source code.

**To Reproduce**
```
use burn::{Tensor, backend::{Autodiff, NdArray, ndarray::NdArrayDevice}, nn::{Linear, LinearConfig}, optim::GradientsParams};

fn main() {
    type MyBackend = NdArray;
    type MyAutodiffBackend = Autodiff<MyBackend>;
    let device = NdArrayDevice::default();
    let linear: Linear<MyAutodiffBackend> = LinearConfig::new(3, 3).init(&device);
    let input: Tensor<MyAutodiffBackend, 1> = Tensor::from_data([1.0, 0.0, -1.0], &device);
    let output = linear.forward(input);
    let loss = output.powi_scalar(2).sum();
    let grads = loss.backward();
    let grads = GradientsParams::from_grads(grads, &linear);
    let weight_grads = grads.get::<MyAutodiffBackend, 2>(linear.weight.id);
    println!("{:?}", weight_grads);
}
```

**Expected behavior**
It should not panic, or at least indicate that while the id was found, it does not correspond to the requested backend.

**Comment by laggui** [comment]
Linking https://github.com/tracel-ai/burn/issues/2924

The reason is simply because gradients are not stored on the autodiff backend. The values in the tensor container for the gradients do not need to be tracked, so they are simply stored on the inner backend.

But I agree that this should be improved. Either via better communication of the expectations, or a more user friendly API.

**Comment by Capataina** [comment]
Both this and #2924 have been open a while (same root cause). VirtualNonsense's PR #2965 got pretty far before going stale. Probably a contained fix, I can patch it up? One quick scoping question. The prior thread had two directions:

- Result-based API on the existing `Box<dyn Any>`
- Make `TensorContainer` generic over Backend (sidesteps the downcast entirely)

I can test both against the reproducer here to verify each closes both threads. Which one seems best? that way I can look into it further

**Comment by laggui** [comment]
Hey @Capataina, thanks for following up.

The main UX issue should be resolved by #4717, which removes the backend generic from the user API. `GradientsParams` will no longer use `B::InnerBackend` to register grads and in turn won't require the correct inner backend generic to be used when retrieving gradients. It should just work.
